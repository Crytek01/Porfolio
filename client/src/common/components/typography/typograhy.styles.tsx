import styled from 'styled-components';

interface ITypoElement {
  $fontSize: string;
  $fontWeight: number;
  $textAlign: string;
  $lineHeight: number;
  $color: string;
}

export const TypoElement = styled.p<ITypoElement>`
  font-size: ${({ $fontSize }) => $fontSize};
  font-weight: ${({ $fontWeight }) => $fontWeight};
  text-align: ${({ $textAlign }) => $textAlign};
  color: ${({ $color }) => $color};
  line-height: ${({ $lineHeight }) => $lineHeight};
`;
