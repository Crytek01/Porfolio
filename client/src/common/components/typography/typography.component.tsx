import { FunctionComponent } from 'react';

import { colors } from '../../../infrastructure/theme/colors';
import { fontSize, fontWeight } from '../../../infrastructure/theme/font';
import { sanitizeHTMLText } from '../../utils/text/text.util';
import { NeverTogether } from '../../utils/typescript/type.util';

import { TypoElement } from './typograhy.styles';

type TElement = {
  p: 'p';
  h1: 'h1';
  h2: 'h2';
  h3: 'h3';
  h4: 'h4';
  h5: 'h5';
  h6: 'h6';
};

export const styleSize = {
  h1: fontSize[7],
  h2: fontSize[6],
  h3: fontSize[5],
  h4: fontSize[4],
  h5: fontSize[3],
  h6: fontSize[2],
  b1: fontSize[5],
  b2: fontSize[4],
  b3: fontSize[3],
  b4: fontSize[2],
  b5: fontSize[1],
  b6: fontSize[0],
};

const styleWeight = {
  regular: fontWeight.regular,
  medium: fontWeight.medium,
  semiBold: fontWeight.semibold,
  bold: fontWeight.bold,
};

type TAlignment = {
  center: 'center';
  left: 'left';
  right: 'right';
  justify: 'justify';
};

interface ITypography {
  as?: keyof TElement;
  size?: keyof typeof styleSize;
  weight?: keyof typeof styleWeight;
  align?: keyof TAlignment;
  color?: keyof typeof colors.text;
  lineHeight?: number;
  className?: string;
  innerHtml?: string;
  children?: React.ReactNode;
}

type TTypography = NeverTogether<ITypography, 'children', 'innerHtml'>;

export const Typography: FunctionComponent<TTypography> = ({
  as = 'p',
  size = 'b3',
  weight = 'regular',
  align = 'unset',
  color = 'standartGrey',
  lineHeight = 1,
  className,
  innerHtml,
  children,
}) => {
  const innerHtmlProp = innerHtml
    ? {
        dangerouslySetInnerHTML: {
          __html: sanitizeHTMLText(innerHtml ?? ''),
        },
      }
    : {};

  return (
    <TypoElement
      className={className}
      as={as}
      fontSize={styleSize[size]}
      fontWeight={styleWeight[weight]}
      textAlign={align}
      color={colors.text[color]}
      lineHeight={lineHeight}
      {...innerHtmlProp}
    >
      {children}
    </TypoElement>
  );
};
