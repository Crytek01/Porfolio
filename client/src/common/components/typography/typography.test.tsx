import { shallowRender } from '@common/utils/test/shallowRender';
import { describe, expect, it } from 'vitest';

import { Typography } from './typography.component';

describe('Typography component Test', () => {
  it('renders correctly', () => {
    const renderer = shallowRender(
      <Typography
        size="b1"
        weight="bold"
        align="center"
        color="boldGrey"
        lineHeight={1}
      >
        Typo
      </Typography>
    );

    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });
});
