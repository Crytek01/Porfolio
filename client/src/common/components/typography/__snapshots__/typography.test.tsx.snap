// Vitest Snapshot v1

exports[`Typography component Test > renders correctly 1`] = `
<styled.p
  as="p"
  color="#333"
  fontSize="2rem"
  fontWeight={700}
  lineHeight={1}
  textAlign="center"
>
  Typo
</styled.p>
`;
